src/test/mon/test_election.cc

```diff
@@ -0,0 +1,369 @@
+	   << ", acked_me:" << logic.acked_me << std::endl;
+      reset_election();
+    }
+    void notify_timestep() {
+      assert(timer_steps != 0);
+      if (timer_steps > 0) {
+	--timer_steps;
+      }
+      if (timer_steps == 0) {
+	if (timer_election) {
+	  election_timeout();
+	} else {
+	  victory_timeout();
+	}
+      }
+    }
+  };
+}
+
+Election::Election(int c) : count(c)
+{
+  for (int i = 0; i < count; ++i) {
+    electors[i] = new Owner(i, this);
+  }
+}
+
+Election::~Election()
+{
+  {
+    for (auto i : electors) {
+      delete i.second;
+    }
+  }
+}
+
+void Election::queue_message(int from, int to, function<void()> m)
+{
+  if (!blocked_messages[from].count(to)) {
+    messages.push_back(m);
+  }
+}
+void Election::defer_to(int from, int to, epoch_t e)
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e] {
+    o->receive_ack(from, e);
+    });
+}
+
+void Election::propose_to(int from, int to, epoch_t e)
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e] {
+      o->receive_propose(from, e);
+    });
+}
+
+void Election::claim_victory(int from, int to, epoch_t e, const set<int>& members)
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e, members] {
+      o->receive_victory_claim(from, e, members);
+    });
+}
+
+void Election::accept_victory(int from, int to, epoch_t e)
+{
+  Owner *o = electors[to];
+  queue_message(from, to, [o, from, e] {
+      o->receive_victory_ack(from, e);
+    });
+}
+
+int Election::run_timesteps(int max)
+{
+  vector< function<void()> > current_m;
+  int steps = 0;
+  for (; (!max || steps < max) && // we have timesteps left AND ONE OF
+	 (!messages.empty() || // there are messages pending.
+	  !election_stable()); // somebody's not happy and will act in future
+       ++steps) {
+    current_m.clear();
+    current_m.swap(messages);
+    for (auto& m : current_m) {
+      m();
+    }
+    for (auto o : electors) {
+      o.second->notify_timestep();
+    }
+  }
+
+  return steps;
+}
+
+void Election::start_one(int who)
+{
+  assert(who < static_cast<int>(electors.size()));
+  electors[who]->logic.start();
+}
+
+void Election::start_all() {
+  for (auto e : electors) {
+    e.second->logic.start();
+  }
+}
+
+bool Election::election_stable()
+{
+  // see if anybody has a timer running
+  for (auto i : electors) {
+    if (i.second->timer_steps != -1)
+      return false;
+  }
+  return true;
+}
+
+void Election::block_messages(int from, int to)
+{
+  blocked_messages[from].insert(to);
+}
+void Election::block_bidirectional_messages(int a, int b)
+{
+  block_messages(a, b);
+  block_messages(b, a);
+}
+void Election::unblock_messages(int from, int to)
+{
+  blocked_messages[from].erase(to);
+}
+void Election::unblock_bidirectional_messages(int a, int b)
+{
+  unblock_messages(a, b);
+  unblock_messages(b, a);
+}
+
+
+TEST(election, binary_runs_at_all)
+{
+  cout << "hello test world!" << std::endl;
```

Comment:
what is this test for?

Reply:
Nothing at all, just leftovers from making sure I was developing it properly. Removed.